tests/baselines/reference/bestCommonTypeOfTuple2.types

```diff
@@ -29,7 +29,7 @@ class F extends C { f }
 class C1 implements base1 { i = "foo"; c }
 >C1 : C1
 >base1 : base1
->i : string
+>i : any

```

Comment:
After more discussion, we decided to drop rule (2) and additionally special-case contextual typing of `null`, `undefined`, `[]` and `{}` so that `undefined/null` in these types takes the contextual type rather than `undefined`.